File: src/FSharpVSPowerTools/Commands/HighlightUsageTaggerProvider.cs
using System.Diagnostics;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.Text.Tagging;
using Microsoft.VisualStudio.Utilities;
using Microsoft.VisualStudio.Shell;
using EnvDTE;
using FSharpVSPowerTools.HighlightUsage;
using FSharpVSPowerTools.ProjectSystem;
using System;

namespace FSharpVSPowerTools
{
    [Export(typeof(IViewTaggerProvider))]
    [ContentType("F#")]
    [TagType(typeof(HighlightUsageTag))]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    [TextViewRole(PredefinedTextViewRoles.PrimaryDocument)]
    public class HighlightUsageTaggerProvider : IViewTaggerProvider
    {
        readonly IServiceProvider _serviceProvider;
        readonly ITextDocumentFactoryService _textDocumentFactoryService;
        readonly ProjectFactory _projectFactory;
        readonly VSLanguageService _fsharpVsLanguageService;

        private readonly IGeneralOptions _generalOptions;
        
        [ImportingConstructor]
        public HighlightUsageTaggerProvider(
            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
            ITextDocumentFactoryService textDocumentFactoryService,
            ProjectFactory projectFactory,
            IGeneralOptions generalOptions,
        
        VSLanguageService fsharpVsLanguageService)
        {
            _serviceProvider = serviceProvider;
            _textDocumentFactoryService = textDocumentFactoryService;
            _projectFactory = projectFactory;
            _fsharpVsLanguageService = fsharpVsLanguageService;
            _generalOptions = generalOptions;
        }

        public ITagger<T> CreateTagger<T>(ITextView textView, ITextBuffer buffer) where T : ITag
        {
            // Only provide highlighting on the top-level buffer
            if (textView.TextBuffer != buffer) return null;

            //var generalOptions = Setting.getGeneralOptions(_serviceProvider);
            if (_generalOptions == null || !_generalOptions.HighlightUsageEnabled) return null;

            ITextDocument doc;
            if (_textDocumentFactoryService.TryGetTextDocument(buffer, out doc))
            {
                return buffer.Properties.GetOrCreateSingletonProperty(
                    () => new HighlightUsageTagger(doc, textView, _fsharpVsLanguageService, _serviceProvider, _projectFactory)) as ITagger<T>;
            }

            return null;
        }
    }
}

File: src/FSharpVSPowerTools.Pure/Commands/HighlightUsageTaggerProvider.fs
module FSharpVSPowerTools.HighlightUsageTaggerProvider

open System.Diagnostics
open System.ComponentModel.Composition
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Operations
open Microsoft.VisualStudio.Text.Tagging
open Microsoft.VisualStudio.Utilities
open Microsoft.VisualStudio.Shell
open EnvDTE
open FSharpVSPowerTools.HighlightUsage
open FSharpVSPowerTools.ProjectSystem
open System

//[<Export(typeof<IViewTaggerProvider>)>]
//[<ContentType("F#")>]
//[<TagType(typeof<HighlightUsageTag>)>]
//type HighlightUsageTaggerProvider [<ImportingConstructor>]
//    ( [<Import(typeof<SVsServiceProvider>)>] 
//        serviceProvider : IServiceProvider,
//        textDocumentFactoryService      :   ITextDocumentFactoryService           ,
//        projectFactory                  :   ProjectFactory                        ,
//        vsLanguageService               :   VSLanguageService                     ) =
//
//    interface IViewTaggerProvider with
//        member __.CreateTagger (textView, buffer) = 
//            maybe{
//          //      let generalOptions = Setting.getGeneralOptions serviceProvider
//                let! doc = textDocumentFactoryService.TryDocumentFromBuffer buffer
//            //    if not generalOptions.HighlightUsageEnabled then return! None else
//                return buffer.Properties.GetOrCreateSingletonProperty (fun () ->
//                    new HighlightUsageTagger (doc, textView, vsLanguageService, serviceProvider, projectFactory)
//                ) :> obj :?> _
//            } |> Option.getOrElse null
